feat(compiler-cli): add `imgAltRequired` extended diagnostic

dgp1130 opened the pull request:
This diagnostic requires that all `<img />` tags have an `alt` attribute with a value. There are a few nuances to this:
* `<img alt />` is considered a mistake which emits a diagnostic.
* By contrast, `<img alt="" />` is considered valid and does _not_ emit a diagnostic because this signals to browsers that the image is decorative and can be ignored for a11y purposes. * See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-alt
* `<img [attr.alt]="altText" />` checks if `altText` is "nullable" (can be `undefined` or `null`). If so, it emits a diagnostic adovising the user to remove those values from the type.

Documentation elaborates on what an `alt` tag is, why you should write one, how to handle each subcase of this diagnostic, and how to turn off the check if desired.

This commit also enables the extended diagnostic as a warning for all Angular projects by default.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Feature

## What is the current behavior?

Issue Number: #46669

`<img />` tags are allowed to not contain an `alt` attribute or contain an `alt` attribute bound to a property which can be `null` or `undefined`. These lead to poor accessibility, as screen readers and other assistive technologies cannot reason about the contents of an image and it is effectively invisible or opaque to many users.

## What is the new behavior?

The Angular compiler emits a diagnostic (configurable severity, warning by default) which alerts the user when they author an `<img />` tag without an `alt` attribute or bind it to a property which can be `null` or `undefined`.

## Does this PR introduce a breaking change?

- [X] No

Like all new extended diagnostics, if users set `defaultCategory: 'error'`, then this change may introduce a new error in their build. However, this is explicitly not supported per our [semver policy for extended diagnostics](https://angular.io/extended-diagnostics#semantic-versioning). This PR is targeting v15 anyways, so even in this situation, a breakage is expected by users.

## Other information

Directives can technically bind `alt` attributes, for example `<img *myImageDirective />` _could_ be valid if `myImageDirective` sets the `alt` attribute. However, this is tricky to statically analyze and uncommon for the `alt` attribute in particular. This particular use cases is ignored for now, and will incorrectly show a warning. https://github.com/angular/angular/issues/47374 would provide more introspection functionality which would allow the compiler to more effectively filter out false positives from such a situation.

Fixes #46669.

dgp1130 committed:
feat(compiler-cli): add `imgAltRequired` extended diagnostic

This diagnostic requires that all `<img />` tags have an `alt` attribute with a value. There are a few nuances to this:
* `<img alt />` is considered a mistake which emits a diagnostic.
* By contrast, `<img alt="" />` is considered valid and does _not_ emit a diagnostic because this signals to browsers that the image is decorative and can be ignored for a11y purposes.
    * See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-alt
* `<img [attr.alt]="altText" />` checks if `altText` is "nullable" (can be `undefined` or `null`). If so, it emits a diagnostic adovising the user to remove those values from the type.

Documentation elaborates on what an `alt` tag is, why you should write one, how to handle each subcase of this diagnostic, and how to turn off the check if desired.

This commit also enables the extended diagnostic as a warning for all Angular projects by default.

dgp1130 commented on aio/content/extended-diagnostics/NG8108.md:
```diff
@@ -0,0 +1,114 @@
+
+</code-example>
+
+## What's wrong with that?
```
@twerske, if you know of any particularly good resources which describe what an `alt` attribute is and how to write a good one, I'm happy to include them. I was surprised by the lack of content on MDN.

I couldn't find any good first-party resources beyond https://www.youtube.com/watch?v=flf2vS0IoRs, but the show is literally named "Non-authoritative information", so I don't know that it's the best reference here.

twerske replied:
That's one of my favorite HTTP203 episodes but yes, it's lengthy and unofficial. 

You're correct, there is a shocking dearth of good reference material. The best resources I can think of would be:
https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Images_in_HTML#alternative_text
https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt
https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML#text_alternatives
https://webaim.org/techniques/alttext/
https://www.w3.org/WAI/tutorials/images/
https://www.w3.org/WAI/tutorials/images/decision-tree/

dgp1130 commented on aio/content/extended-diagnostics/NG8108.md:
```diff
@@ -0,0 +1,114 @@
+</code-example>
+
+If the image is used purely for visual styling with no semantic meaning (like a spacer), then set
+`alt=""`, which tells browsers to ignore the image for accessibility purposes.
```
@twerske, it's my understanding that `alt=""` is occasionally correct for semantically meaningless images, but `alt` is _never_ correct. Let me know if I'm mistaken and I can update the logic / documentation.

twerske replied:
yes this is correct (source: https://www.w3.org/WAI/tutorials/images/decorative/). I wonder if it's worth having a warning or language service message on `alt=""` that says "This is only ok if this is a purely decorative image"

dgp1130 replied:
I'd rather avoid unactionable messages so devs aren't incentivized to turn off the check. If a dev writes `alt=""` then either the message is purely decorative or they're deliberately trying to silence the warning and won't be receptive to the message anyways.

twerske replied:
Makes sense - resolving this comment thread!

jelbourn commented:
> <img [attr.alt]="altText" /> checks if altText is "nullable" (can be undefined or null). If so, it emits a diagnostic adovising the user to remove those values from the type.

Is this going to cause a problem in situations where the alt text comes from an async request and is `null` or `undefined` until the request resolves?

dgp1130 commented:
> > <img [attr.alt]="altText" /> checks if altText is "nullable" (can be undefined or null). If so, it emits a diagnostic adovising the user to remove those values from the type.
> 
> Is this going to cause a problem in situations where the alt text comes from an async request and is `null` or `undefined` until the request resolves?

Possibly. If you do something like `<img [attr.alt]="altText$ | async" />` it would probably warn. But also that is a real issue since you don't have alt text for part of the image's lifespan. I think the appropriate fix would be to `*ngIf` the whole `<img />` tag so it's only rendered when you definitely have alt text. If you're dynamically fetching the `src` attribute and getting `alt` at the same time, I think that would mostly "just work" and should help drive people towards the right solution.

dgp1130 commented on aio/content/extended-diagnostics/NG8108.md:
```diff
@@ -0,0 +1,117 @@
+
+@description
+
+This diagnostic detects an `<img />` tag missing an
```
I personally have that style, but I'm not sure what we've typically done for AIO. Happy to adjust it to whatever style we decide.

twerske replied:
@bob-watson for future PRs I think this is a good inclusion in the linter for consistency across docs. 

Here I think the `/` is fine though

twerske reviewed (approved):
LGTM from DevRel - thank you for working on this @dgp1130 - this is really revolutionary in the scope of building-in a11y to the developer experience on the web!!!

dgp1130 committed:
refactor: bump components SHA to latest `main` commit

This updates the components CI test to use the latest commit on their `main` branch (https://github.com/angular/components/commit/a999a52bb967608b4ec59729cf720ade9db274d8). This is needed to test with https://github.com/angular/components/pull/25911, which fixes an issue introduced by the new extended diagnostic.

josephperrott reviewed (approved):
LGTM

JeanMeche commented on packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts:
```diff
@@ -318,6 +318,11 @@ export enum ErrorCode {
+  /**
+   * `<img />` tags require the `alt` attribute to provide accessibility information.
+   */
+  IMG_ALT_REQUIRED = 8108,
```
`8108` is not available anymore, you'll need to rebase your PR !

dgp1130 commented:
Switching this to draft since it was stuck in review for long enough that it's run into a number of merge conflicts.

IIRC it was also breaking the components repo because they configured extended diagnostics to error by default, so we'd need to resolve that as well.
